File: external/java/src/main/java/ch/epfl/dedis/lib/crypto/PointFactory.java
package ch.epfl.dedis.lib.crypto;

import ch.epfl.dedis.lib.exception.CothorityCryptoException;
import com.google.protobuf.ByteString;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * A factory to create point from different input. A protobuf input will have a tag at the beginning
 * to let the factory know which point to create and a point coming from a toml file will require the
 * suite information that must be available.
 */
public class PointFactory {
    static public final String SUITE_ED25519 = "Ed25519";
    static public final String SUITE_BN256 = "Bn256.adapter";

    static private final Logger logger = LoggerFactory.getLogger(Ed25519Point.class);
    static private final PointFactory INSTANCE = new PointFactory();

    /**
     * Get the singleton instance of the factory
     * @return the instance
     */
    static public PointFactory getInstance() {
        return INSTANCE;
    }

    private Map<Tag, PointGenerator> tags;

    /**
     * Instantiate the factory with bindings to Ed25519 and Bn256 points. More must be
     * added for new kind of points.
     */
    private PointFactory() {
        tags = new HashMap<>();
        tags.put(new Tag(Bn256G2Point.marshalID), Bn256G2Point::new);
        tags.put(new Tag(Ed25519Point.marshalID), Ed25519Point::new);
    }

    /**
     * Create a point using a protobuf input
     * @param pubkey the byte string of the public key
     * @return the point or null
     */
    public Point fromProto(ByteString pubkey) {
        byte[] buf = Arrays.copyOfRange(pubkey.toByteArray(), 0, 8);
        Tag tag = new Tag(buf);

        if (tags.containsKey(tag)) {
            try {
                return tags.get(tag).make(pubkey.toByteArray());
            } catch (CothorityCryptoException e) {
                logger.error(e.getMessage());
            }
        }

        return null;
    }

    /**
     * Create a point using a toml input string
     * @param suite the name of the suite to use to create the point
     * @param pubhex the hex string of the public key
     * @return the point or null
     */
    public Point fromToml(String suite, String pubhex) {
        try {
            switch (suite) {
                case PointFactory.SUITE_ED25519:
                    return new Ed25519Point(pubhex);
                case PointFactory.SUITE_BN256:
                    return new Bn256G2Point(pubhex);
                default:
                    return null;
            }
        } catch (CothorityCryptoException e) {
            logger.error(e.getMessage());
        }

        return null;
    }

    /**
     * Generator used to create point of the right suite for given tags
     */
    private interface PointGenerator {
        /**
         * Instantiate the point
         * @param data byte array of the point with the tag still prepended
         * @return the point
         */
        Point make(byte[] data) throws CothorityCryptoException;
    }

    /**
     * Tag used as a key to bind a generator to a point type
     */
    private class Tag {
        private byte[] tag;

        Tag(byte[] value) {
            tag = value;
        }

        @Override
        public boolean equals(Object other) {
            if (!(other instanceof Tag)) {
                return false;
            }

            return Arrays.equals(tag, ((Tag) other).tag);
        }

        @Override
        public int hashCode() {
            return ByteBuffer.wrap(tag).getInt();
        }
    }
}

File: external/js/kyber/src/point-factory.ts
import { Point } from ".";
import { BN256G1Point, BN256G2Point } from "./pairing/point";
import Ed25519Point from "./curve/edwards25519/point";

interface GeneratorMap {
    [k: string]: () => Point
}

/**
 * Factory that must be used to decode points coming from the network or
 * a TOML file. It will take care of looking up the right instance of the
 * point. A dedicated toProto function is provided for the points to encode
 * them back if required.
 */
export class PointFactory {
    private static SUITE_ED25519 = 'Ed25519';
    private static SUITE_BN256 = 'bn256.adapter';

    private tags: GeneratorMap;
    private suites: GeneratorMap;

    constructor() {
        this.tags = {
            [`${BN256G1Point.MARSHAL_ID.toString()}`]: () => new BN256G1Point(),
            [`${BN256G2Point.MARSHAL_ID.toString()}`]: () => new BN256G2Point(),
            [`${Ed25519Point.MARSHAL_ID.toString()}`]: () => new Ed25519Point(),
        };

        this.suites = {
            [`${PointFactory.SUITE_ED25519}`]: () => new Ed25519Point(),
            [`${PointFactory.SUITE_BN256}`]: () => new BN256G2Point(),
        };
    }

    /**
     * Decode a point using the 8 first bytes to look up the
     * correct instance
     * 
     * @param buf bytes to decode
     * @returns the point
     */
    public fromProto(buf: Buffer): Point {
        const tag = buf.slice(0, 8).toString();
        const generator = this.tags[tag];

        if (generator) {
            const point = generator();
            point.unmarshalBinary(buf.slice(8));

            return point
        }

        throw new Error('unknown tag for the point');
    }

    /**
     * Decode the point stored as an hexadecimal string by using
     * the suite as a reference for the point instance
     * 
     * @param suite Name of the suite to use
     * @param pub   The point encoded as an hex-string
     * @returns the point
     */
    public fromToml(suite: string, pub: string): Point {
        const generator = this.suites[suite];

        if (generator) {
            const point = generator();
            const buf = Buffer.from(pub, 'hex');
            point.unmarshalBinary(buf);

            return point;
        }

        throw new Error('unknown suite for the point');
    }
}

export default new PointFactory();
